nixos/direnv: init

Gerg-L opened the pull request:
Closes #192667

###### Description of changes
There really should be a simple direnv module and #192667 has become quite stale
So I took most of the criticisms and applied them along with a few bits of my own

I would like to make an option to conditionally load based on being in a nix shell, 
but alas i couldn't get `fish` to work correctly

###### Things done
- [x] Tested the module

Gerg-L committed:
nixos/direnv: init

drupol commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,88 @@
+      bash.interactiveShellInit = ''
+        eval "$(direnv hook bash)"
+      '';
+      fish.interactiveShellInit = ''
```
What if there is already such an attribute with a custom value? Is this going to override it ?

Gerg-L replied:
It's of type "list of lines" so it'll be merged

drupol replied:
Cool ! Looking forward for this thing!

drupol reviewed (approved):
LGTM

2xsaiko commented:
FWIW I have some extra code in my config's module for direnv to source the user's configuration also (which direnv normally does), might be worth including here. This is what configuration I install (in /etc/xdg/direnv as opposed to /etc/direnv):

```
    environment.etc = {
      "xdg/direnv/lib/nix-direnv.sh".source = mkIf cfg.enableNixDirenv "${pkgs.nix-direnv}/share/nix-direnv/direnvrc";

      # Also load user's config files
      "xdg/direnv/lib/zz-user.sh".text = ''
        direnv_config_dir_home="''${DIRENV_CONFIG_HOME:-''${XDG_CONFIG_HOME:-$HOME/.config}/direnv}"

        for lib in "$direnv_config_dir_home/lib/"*.sh; do
          source "$lib"
        done

        unset direnv_config_dir_home
      '';

      "xdg/direnv/direnvrc".text = ''
        direnv_config_dir_home="''${DIRENV_CONFIG_HOME:-''${XDG_CONFIG_HOME:-$HOME/.config}/direnv}"

        # load the global ~/.direnvrc if present
        if [[ -f $direnv_config_dir_home/direnvrc ]]; then
          # shellcheck disable=SC1090,SC1091
          source "direnv_config_dir_home/direnvrc" >&2
        elif [[ -f $HOME/.direnvrc ]]; then
          # shellcheck disable=SC1090,SC1091
          source "$HOME/.direnvrc" >&2
        fi

        unset direnv_config_dir_home
      '';
    };
```

Gerg-L commented:
> FWIW I have some extra code in my config's module for direnv to source the user's configuration also (which direnv normally does), might be worth including here. 

I can't see any negatives to this I'll add it

2xsaiko commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,109 @@
+          fi
+
+          unset direnv_config_dir_home
+            ${cfg.direnvrcExtra}
```
Running extra config here should probably be done before loading user config

Gerg-L replied:
Fixed

SuperSandro2000 commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,112 @@
+      type = lib.types.lines;
+      default = "";
+      example = ''
+        use_zig() {
+          zig_inst -q $1
+          ZIG_DIR="''${HOME}/.local/share/zig/$1"
+          ZIG_UPDATE="''${ZIG_DIR}/zig_update"
+          if [ ! -f "''${ZIG_UPDATE}" ]; then
+            echo -e "#!/bin/bash\nzig_inst -q -f $1\n" > "''${ZIG_UPDATE}"
+           chmod a+x "''${ZIG_UPDATE}"
+          fi
+          PATH_add "''${ZIG_DIR}"
+          PATH_add "./zig-cache/bin"
+          PATH_add "./zig-out/bin"
+          echo "Using zig $1"
+        }
```
Can we use a generic example which can't possible be wrong in the details?

Gerg-L replied:
Done

SuperSandro2000 commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,112 @@
+    };
+  };
+  config = lib.mkIf cfg.enable {
+    programs = {
+      zsh.interactiveShellInit = ''
+        eval "$(direnv hook zsh)"
+      '';
+      bash.interactiveShellInit = ''
+        eval "$(direnv hook bash)"
+      '';
+      fish.interactiveShellInit = ''
+        direnv hook fish | source;
+      '';
+    };
+    nix.settings = lib.mkIf cfg.persistDerivations {
+      keep-outputs = true;
+      keep-derivations = true;
+    };
+    environment = {
```
```suggestion
    programs = {
      zsh.interactiveShellInit = ''
        eval "$(direnv hook zsh)"
      '';
      bash.interactiveShellInit = ''
        eval "$(direnv hook bash)"
      '';
      fish.interactiveShellInit = ''
        direnv hook fish | source;
      '';
    };

    nix.settings = lib.mkIf cfg.persistDerivations {
      keep-outputs = true;
      keep-derivations = true;
    };

    environment = {
```
having some new lines improves readability a lot

SuperSandro2000 commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,112 @@
+      keep-derivations = true;
+    };
+    environment = {
+      systemPackages = [pkgs.direnv];
```
```suggestion
      systemPackages = with pkgs; [ direnv ];
```
don't we need nix-direnv? https://github.com/NixOS/nixpkgs/pull/192667/files#diff-ff3b3a796d514f9f77a780213ec8446daf34394c6b097f837cc226231b21af30R16-R19

Gerg-L replied:
No need with this strategy it can just be sourced directly in the direnvrc

Gerg-L replied:
Oh! I accidentally deleted the sourcing on the last push

bbenne10 replied:
To be clear: It's already in the direnvrc. We are referencing it in `/etc/direnv/direnvrc` (see 108 below)

fufexan commented:
Missclick, sorry!

bbenne10 commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,136 @@
+          #Load direnvrcExtra
+          ${cfg.direnvrcExtra}
+
+          #Load ~/.direnvrc if present
```
This is a bit confusing. Maybe something like this? 
```diff
- #Load ~/.direnvrc if present
+ #Load user-configuration if present (~/.direnvrc or ~/.config/direnv/direnvrc)
```

Gerg-L replied:
Done!

bbenne10 reviewed (approved):
Other than one (admittedly rather nitpicky) concern re a comment: this looks like a very good approach that addresses every problem I have with #192667!

Gerg-L commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,136 @@
+  options.programs.direnv = {
+
+    enable = lib.mkOption {
+
```
https://github.com/NixOS/nixpkgs/pull/241528#discussion_r1254324710

drupol replied:
I also agree but there's no need to add an empty between _each_ lines.

Can you please remove the unnecessary empty lines and mimic what's in this file ?

https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/ssh.nix#L36

drupol replied:
Definitely, thanks !

jian-lin commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,136 @@
+        echo "loaded direnv!"
+      '';
+
+      description = ''
```
Use `lib.mdDoc` and it will be much easier to backport this module to 23.05.

jian-lin commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,136 @@
+
+  options.programs.direnv = {
+
+    enable = lib.mkOption {
```
The is a util function `mkEnableOption` for bool option.

Example: https://github.com/NixOS/nixpkgs/blob/873012e0ac6c4feb66eeb88d7e14df4a8323f565/nixos/modules/services/hardware/kanata.nix#L142

Gerg-L replied:
Added for enable

jian-lin commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,111 @@
+    direnvrcExtra = lib.mkOption {
+      type = lib.types.lines;
+      default = "";
+      example = lib.literalExpression ''
```
No need to add this, see https://nixos.org/manual/nixos/unstable/#sec-option-declarations
```suggestion
      example = ''
```

jian-lin commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,111 @@
+  options.programs.direnv = {
+
+    enable = lib.mkEnableOption (lib.mdDoc ''
+      Whether to enable direnv integration. Takes care of both installation and
```
mkEnableOption adds `Whether to enable ` to the beginning of your string for you.

jian-lin commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,111 @@
+      '';
+    };
+
+    persistDerivations = lib.mkOption {
```
`mkEnableOption` can also be used here.

Gerg-L replied:
How do set `mkEnableOption` to default to false?

Gerg-L replied:
Yes thank you

jian-lin commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,111 @@
+      '';
+    };
+
+    silent = lib.mkOption {
```
mkEnableOption can also be used here.

drupol commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,111 @@
+
+    environment = {
+
+      systemPackages = [pkgs.direnv];
```
Shouldn't we add something like :       

```nix
package = lib.mkPackageOptionMD pkgs "direnv" { };
```

In the configuration?

drupol reviewed (commented):
Now that you added a `package` attribute in the config, you can use it as such...

Gerg-L commented:
> Now that you added a `package` attribute in the config, you can use it as such... 

Does the package gets added to automatically?
Probably not right?

drupol commented:
> Does the package gets added to automatically to systemPackages? Probably not right?

No it doesn't but your PR doesn't do it either.

Gerg-L commented:
> > Does the package gets added to automatically to systemPackages? Probably not right?
> 
> No it doesn't but your PR doesn't do it either.

Yeah removed it with the last force push, brain fart

drupol commented:
> Yeah removed it with the last force push, brain fart

Then you just have to use `cfg.package`.

Gerg-L commented:
Alright pushed changes

drupol commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,106 @@
+      etc = {
+        "direnv/direnvrc".text = ''
+          #Load nix-direnv
+          source ${pkgs.nix-direnv}/share/nix-direnv/direnvrc
```
Maybe we should also let `nix-direnv` be configured in the options ? WDYT?

Something like:

```nix
package-nix-direnv = lib.mkPackageOptionMD pkgs "nix-direnv" {};
```

Gerg-L replied:
Finding the right path to source wouldn't be easy... If anything I'd remove the sourcing of `nix-direnv` and say use `direnvrcExtraConfig`

Gerg-L commented:
I finally figured out why fish wasn't working with my conditional loading setup so I added a loadInNixShell option, I don't know whether this should be an override to `direnv` or a change to the package itself it doesn't seem right that it automatically is sourcing itself when installed

I also added options to disable `nix-direnv` as drupol suggested

drupol commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,148 @@
+    };
+
+    environment = {
+      systemPackages =
+        #direnv has a fish library which sources direnv for some reason
+        if cfg.loadInNixShell
+        then [cfg.package]
+        else [
+          (cfg.package.overrideAttrs (old: {
+            installPhase =
+              (old.installPhase or "")
+              + ''
+                rm -rf $out/share/fish
+              '';
+          }))
+        ];
```
```suggestion
      systemPackages =
        if cfg.loadInNixShell then 
          [cfg.package]
        else [
          #direnv has a fish library which sources direnv for some reason
          (cfg.package.overrideAttrs (old: {
            installPhase =
              (old.installPhase or "")
              + ''
                rm -rf $out/share/fish
              '';
          }))
        ];
```

drupol reviewed (changes requested):
A minor syntax change and I think it's good to go.

Gerg-L commented:
> A minor syntax change and I think it's good to go.

Changes pushed!

drupol commented on nixos/modules/programs/direnv.nix:
```diff
@@ -0,0 +1,147 @@
+
+    environment = {
+      systemPackages =
+        if cfg.loadInNixShell then [cfg.package]
```
Ok, I'm not going to do bikeshedding here, but the syntax is not the most optimal. Let's leave it like that for the moment.

Gerg-L replied:
Oops...
